### PR TITLE
Change to use multiple logs in Rust training course

### LINKED ISSUE #71: Use multiple logs in Rust training course code
In the text of project 2, we intended to simplify the bitcask algorithm using only one log file. After some practice, it is found that a single log file makes implementation more complex instead of simplifying it.

So I think we should not ask students to use only one log, and then give instructions of using multiple logs or at least give some hints.

### PR DESCRIPTION
**Pull request by sticnarf**
Try to fix #71. You can see lots of subtle stuff is removed thanks to multiple logs.

Before modifying the code in subsequent projects, I want to hear from you @brson first. If you think it's ok, I will add commits to change the code in Project 3.

### CODE REVIEW

**Comment by brson** [comment]
I agree this code looks quite gorgeous. I am not clear though on all the reasons _why_ this is simpler than the single log file. Is it mostly because in future concurrent projects it will be relatively trivial to continue writing while compacting?

I do see that during compaction the index is rebuilt in place, where the old code had to create a new index. I've forgotten what the limitation was that forced the old code to do that though.

Can you summarize in a few sentences the simplifications this approach enables? Please find somewhere in the code to comment something similar as well.

It does strike me that the project is getting quite sophisticated, and this is still project 2. We'll have to think about how to present this in the text. E.g. it's not _necessary_, particularly in project 1 that a student arrive at a multiple-log solution, but as the projects progress we're definitely going to need to indicate the problems solved by multiple logs and steer them to that solution, if not outright insist they use multiple logs.

Please continue this for the subsequent projects.

**Comment by sticnarf** [comment]
The reason why this is simpler is the difficulty to implement correctly using a single log. Multiple logs is a more natural way as long as we support compaction.

The root of the problem is that single log is a limitation, and compaction itself breaks it. During the compaction, there must be at least two log files, the old one, and the new one you write merged records to. But if we can only allow one valid log file, we must use either the old one or the new one, the state has to *jump* from the old one to the new one. It's really subtle to handle the jumping process well (keep it correct and not sacrifice performance much). Of course, it is easy and correct to re-read the new log after the compaction, but unaccepted in performance. I believe many students won't accept this way either, but they will find it difficult to get things correct.

However, with multiple logs, only the file index increases the complexity and it is natural to have that.

So, I don't know how to state why the multiple logs way is simpler in comments. You won't find it actually more difficult until you try the single log way yourself. If I must summarize, compaction itself uses multiple logs, so using multiple logs is natural. That's all.

**Comment by sticnarf** [comment]
Project 3 has switched to the new implementation. You can review it.

Note that the receiver specs of the KvsEngine trait functions are modified in Project 3.

**Comment by brson** [comment]
lgtm
